Fork safe ConcurrentMultiSpanProcessor

gregoiredx opened the pull request:
# Description

<!--
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.
-->

`ConcurrentMultiSpanProcessor` is not "fork safe". This MR tries to solve this problem.

Forking a threaded process is not a good idea in the first place, citing the Python doc, see https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods
> Note that safely forking a multithreaded process is problematic.

Also, the OTel Python SDK suggest to only initialize the SDK post fork, see https://opentelemetry-python.readthedocs.io/en/latest/examples/fork-process-model/README.html

Nevertheless, on an existing code base, it's not always easy to avoid all forks. So it looks like a good idea to try and make the SDK as fork safe as possible.

Note: `BatchProcessor` is already relying on `register_at_fork` used here, see https://github.com/open-telemetry/opentelemetry-python/blob/102fec2be1fe9d0a8e299598a21ad6ec3b96dcca/opentelemetry-sdk/src/opentelemetry/sdk/_shared_internal/__init__.py#L125

Fixes https://github.com/open-telemetry/opentelemetry-python/issues/2349

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

See unit test and discussions below.

An equivalent patch is currently running in our stack.

# Does This PR Require a Contrib Repo Change?

?

# Checklist:

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

DylanRussell commented on opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py:
```diff
@@ -238,8 +239,16 @@ def __init__(self, num_threads: int = 2):
+            os.register_at_fork(after_in_child=self._build_executor)
+
+    def _build_executor(self) -> concurrent.futures.ThreadPoolExecutor:
+        return concurrent.futures.ThreadPoolExecutor(
```
Is there something that says explicitly THreadPoolExecutor needs to be reinitialized in forked processes ? I see the note about not forking threaded processes, but it's still a bit unclear to me..

gregoiredx replied:
More on this, specifically on Linux, https://man7.org/linux/man-pages/man2/fork.2.html
> The child process is created with a single thread—the one that called fork().  The entire virtual address space of the parent is replicated in the child, including the states of mutexes, condition variables, and other pthreads objects; the use of [pthread_atfork(3)](https://man7.org/linux/man-pages/man3/pthread_atfork.3.html) may be helpful for dealing with problems that this can cause.

gregoiredx replied:
I still don't see how to add fork to the actual test suite but here is at least an example I tried to put up to help test what's going on here:
```python
import os
from concurrent.futures import ThreadPoolExecutor

executor = ThreadPoolExecutor(max_workers=2)

def reinit_executor():
    global executor
    executor = ThreadPoolExecutor(max_workers=2)

# Remove this line to see the child process hanging forever:
os.register_at_fork(after_in_child=reinit_executor)

print(f"pid:{os.getpid()} in main thread before fork")
print(executor.submit(lambda : f"pid:{os.getpid()} in sub thread before fork").result())

child_pid = os.fork()
if child_pid:
    print(f"pid:{os.getpid()} forked parent process to child process", child_pid)
else:
    print(f"pid:{os.getpid()} forked process, in child")

print(f"pid:{os.getpid()} in main thread after fork")
print(executor.submit(lambda: f"pid:{os.getpid()} in sub thread after fork").result())

if child_pid:
    print(f"pid:{os.getpid()} waiting for child process")
    os.waitpid(child_pid, 0)
else:
    print(f"pid:{os.getpid()} child process end")
```

gregoiredx commented on opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py:
```diff
@@ -238,8 +239,16 @@ def __init__(self, num_threads: int = 2):
+        if hasattr(os, "register_at_fork"):
+            # Only the main thread is kept in forked processed, the executor
+            # needs to be re-instantiated to get a fresh pool of threads:
+            os.register_at_fork(after_in_child=self._build_executor)
```
Very well spotted! A weakref was indeed needed

gregoiredx commented:
@DylanRussell thanks a lot for the review.
The first pass of the CI wasn't successful.
I applied two fixes:
- `multiprocessing.set_start_method` was already started in another test and it can be called only once, see https://github.com/open-telemetry/opentelemetry-python/actions/runs/20662185206/job/59826482487#step:5:778. I fixed it by switching to `multiprocessing.get_context`
- changelog was missing, I added a line to CHANGELOG.md

There's another test which failed but it doesn't look like it's related to my change, see https://github.com/open-telemetry/opentelemetry-python/actions/runs/20662185256/job/59826483790#step:6:39

Can you relaunch the CI please?
